src/gen.py

```diff
@@ -0,0 +1,363 @@
+        tests.append(
+            tmpl.format(
+                name=func_name,
+                module=csharp_str(name),
+                line=i + 1,
+                col=col,
+                args=csharp_str(args),
+                filter=csharp_str(filt),
+            )
+        )
+
+    if tests:
+        print(CLASS_PREAMBLE.format(name=camel_name))
+        for t in tests:
+            print(t)
+        print(CLASS_POSTAMBLE)
+
+
+DEFAULT_TEST_FILES = [
+    "arrays",
+    "async_",
+    "basic",
+    "classes",
+    "completion",
+    "complex",
+    "comprehensions",
+    "context",
+    "decorators",
+    "definition",
+    "descriptors",
+    "docstring",
+    "dynamic_arrays",
+    "dynamic_params",
+    "flow_analysis",
+    "fstring",
+    "functions",
+    "generators",
+    "imports",
+    "invalid",
+    "isinstance",
+    "keywords",
+    "lambdas",
+    "named_param",
+    "on_import",
+    "ordering",
+    "parser",
+    "pep0484_basic",
+    "pep0484_comments",
+    "pep0484_typing",
+    "pep0526_variables",
+    "precedence",
+    "recursion",
+    "stdlib",
+    "sys_path",
+    "types",
+]
+
+PREAMBLE = """// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using AnalysisTests;
+using FluentAssertions;
+using Microsoft.Python.LanguageServer.Implementation;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Analysis.FluentAssertions;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using TestUtilities;
+
+namespace GenTests {"""
+
+POSTAMBLE = """
+    public class GenTest : ServerBasedTest {
+        private static Server _server;
+        private static readonly object _lock = new object();
+        private static readonly InterpreterConfiguration _interpreter = PythonVersions.LatestAvailable3X;
+        private static readonly PythonLanguageVersion _version = _interpreter.Version.ToLanguageVersion();
+        private static readonly ConcurrentDictionary<string, Task> _opened = new ConcurrentDictionary<string, Task>();
+
+        private Server SharedServer {
```

Comment:
This is called from async context, so it will create async over sync over async calls. You can use async method instead of property.

Reply:
If I do this, then I have to abandon locking because you cannot use `lock` inside of an async function.

Yes, `lock` is overkill because we don't actually run tests in parallel, but that's something I'd like to try to get working in the future.

Reply:
`await SemaphoreSlim.WaitAsync` is your friend

Reply:
Ah, yes, I can do that.